Graphics/Implicit/Export/DiscreteAproxable.hs

```diff
@@ -44,86 +35,101 @@ unmesh (TriangleMesh m) = m
+                0.5
+                box
+                scene
+            ]
         where
-            -- | Size of the image to produce.
-            (V2 w h) = V2 150 150 :: ℝ2
-            obj = getImplicit symbObj
-            box@(V3 x1 y1 z1, V3 _ y2 z2) = getBox3 symbObj
-            av :: ℝ -> ℝ -> ℝ
-            av a b = (a+b)/2
-            avY = av y1 y2
-            avZ = av z1 z2
-            deviation = maximum [abs $ y1 - avY, abs $ y2 - avY, abs $ z1 - avZ, abs $ z2 - avZ]
-            camera = Camera (V3 (x1-deviation*2.2) avY avZ) (V3 0 (-1) 0) (V3 0 0 (-1)) 1.0
-            lights = [Light (V3 (x1-deviation*1.5) (y1 - 0.4*(y2-y1)) avZ) (0.03*deviation) ]
-            scene = Scene obj (Color 200 200 230 255) lights (Color 255 255 255 0)
-            -- | passed to generateImage, it's external, and determines this type.
-            pixelRenderer :: Int -> Int -> PixelRGBA8
-            pixelRenderer a b = renderScreen
-                (fromIntegral a/w - 0.5) (fromIntegral b/h - 0.5)
-            renderScreen :: ℝ -> ℝ -> PixelRGBA8
-            renderScreen a b =
-                colorToPixelRGBA8 $
-                    average [
-                        traceRay
-                            (cameraRay camera (V2 a b + V2 ( 0.25/w) (0.25/h)))
-                            2 box scene,
-                        traceRay
-                            (cameraRay camera (V2 a b + V2 (-0.25/w) (0.25/h)))
-                            0.5 box scene,
-                        traceRay
-                            (cameraRay camera (V2 a b + V2 (0.25/w) (-0.25/h)))
-                            0.5 box scene,
-                        traceRay
-                            (cameraRay camera (V2 a b + V2 (-0.25/w) (-0.25/h)))
-                            0.5 box scene
-                        ]
-                    where
-                      colorToPixelRGBA8 :: Color -> PixelRGBA8
-                      colorToPixelRGBA8 (Color rr gg bb aa) = PixelRGBA8 rr gg bb aa
+          colorToPixelRGBA8 :: Color -> PixelRGBA8
+          colorToPixelRGBA8 (Color rr gg bb aa) = PixelRGBA8 rr gg bb aa
 
 instance DiscreteAproxable SymbolicObj2 [Polyline] where
-    discreteAprox res = symbolicGetContour res defaultObjectContext
+  discreteAprox res = symbolicGetContour res defaultObjectContext
 
 -- FIXME: way too many magic numbers.
 -- FIXME: adjustable resolution?
 instance DiscreteAproxable SymbolicObj2 DynamicImage where
-    discreteAprox _ symbObj = ImageRGBA8 $ generateImage pixelRenderer (round w) (round h)
+  discreteAprox _ symbObj = ImageRGBA8 $ generateImage pixelRenderer (round w) (round h)
+    where
+      V2 w h = pure 150 :: ℝ2
+      obj = getImplicit symbObj
+      (p1@(V2 x1 _), p2@(V2 _ y2)) = getBox2 symbObj
+      V2 dx dy = p2 - p1
+      dxy = max dx dy
+
+      pixelRenderer :: Int -> Int -> PixelRGBA8
+      pixelRenderer mya myb = mycolor
         where
-            -- | Size of the image to produce.
```

Comment:
bad smell, but where did this comment go?

Reply:
hehe: 755d8aa

Reply:
We're finding good stuff already!